### PR TITLE
e4s external rocm ci: bump rocm stack to v6.1.1

### PR DESCRIPTION
**Pull request by eugeneswalker**
E4S External ROCm CI: Use latest ROCm 6.1.1

### CODE REVIEW

**Review by tldahlgren — DISMISSED** [review]
LGTM

**Comment by eugeneswalker** [comment]
@spackbot run pipeline

**Comment by tldahlgren** [comment]
Looks like the rocm external stack is failing CI for a couple of packages (esp. `gasnet`).

**Comment by bonachea** [comment]
> Looks like the rocm external stack is failing CI for a couple of packages (esp. `gasnet`).

GASNet 2023.9.0 has a known incompatibility with ROCm 6 ([bug 4686](https://gasnet-bugs.lbl.gov/bugzilla/show_bug.cgi?id=4686)), which is fixed in the latest GASNet 2024.5.0 release and `master` branch.

**Comment by bonachea** [comment]
> Looks like the rocm external stack is failing CI for a couple of packages (esp. `gasnet`).

@tldahlgren see #44478

**Comment by bonachea** [comment]
@tldahlgren PR #44478 is now merged to spack:develop, so rebasing this PR onto the current [develop](spack:develop) should allow the `gasnet` tests in the failing pipeline to pass.

The failing `upcxx` tests have the same root cause, and will similarly be resolved once PR #44488 has been merged.

**Comment by eugeneswalker** [comment]
`gasnet@2024.5.0 +rocm` failing after rebase because some of the properties referenced in package.py don't exist for externally defined components:
* [ gasnet@2024.5.0 /mmkqcp7 %gcc@11.4.0 arch=linux-ubuntu22.04-x86_64_v3 E4S ROCm External ](https://gitlab.spack.io/spack/spack/-/jobs/11464763) 

```
...
==> Installing gasnet-2024.5.0-mmkqcp7kqr3uq2cogvqgrmb3dvhaw4jz [1/1]
==> Error: KeyError: 'No spec with name hsa-rocr-dev in hip@=6.1.1%gcc@=11.4.0~cuda+rocm build_system=cmake build_type=Release generator=make patches=1f65dfe,6bbe83b arch=linux-ubuntu22.04-x86_64_v3'
/builds/spack/spack/var/spack/repos/builtin/packages/hip/package.py:416, in set_variables:
        413            if self.spec.satisfies("@5.4:"):
        414                env.set("HIPIFY_CLANG_PATH", paths["hipify-clang"])
        415            if self.spec.satisfies("@6.1:"):
  >>    416                env.prepend_path("LD_LIBRARY_PATH", self.spec["hsa-rocr-dev"].prefix.lib)
        417
        418            # hipcc recognizes HIP_PLATFORM == hcc and HIP_COMPILER == clang, even
        419            # though below we specified HIP_PLATFORM=rocclr and HIP_COMPILER=clang
Traceback (most recent call last):
  File "/builds/spack/spack/lib/spack/spack/spec.py", line 4177, in __getitem__
    child: Spec = next(
StopIteration
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/builds/spack/spack/lib/spack/spack/build_environment.py", line 1133, in _setup_pkg_and_run
    kwargs["env_modifications"] = setup_package(
  File "/builds/spack/spack/lib/spack/spack/build_environment.py", line 792, in setup_package
    env_mods.extend(setup_context.get_env_modifications())
  File "/builds/spack/spack/lib/spack/spack/build_environment.py", line 1042, in get_env_modifications
    spack.builder.create(pkg).setup_dependent_build_environment(env, root)
  File "/builds/spack/spack/lib/spack/spack/builder.py", line 202, in setup_dependent_build_environment
    return self.pkg_with_dispatcher.setup_dependent_build_environment(env, dependent_spec)
  File "/builds/spack/spack/var/spack/repos/builtin/packages/hip/package.py", line 481, in setup_dependent_build_environment
  File "/builds/spack/spack/var/spack/repos/builtin/packages/hip/package.py", line 417, in set_variables
  File "/builds/spack/spack/lib/spack/spack/spec.py", line 4191, in __getitem__
    raise KeyError(f"No spec with name {name} in {self}")
KeyError: 'No spec with name hsa-rocr-dev in hip@=6.1.1%gcc@=11.4.0~cuda+rocm build_system=cmake build_type=Release generator=make patches=1f65dfe,6bbe83b arch=linux-ubuntu22.04-x86_64_v3'
```

This has come up in other contexts with other packages where dependencies are externally defined. Will have to go back and see what the right way to handle these cases is.

@bonachea

**Comment by bonachea** [comment]
> `gasnet@2024.5.0 +rocm` failing after rebase because some of the properties referenced in package.py don't exist for externally defined components:
> ...
> This has come up in other contexts with other packages where dependencies are externally defined. Will have to go back and see what the right way to handle these cases is.

@eugeneswalker IIUC the error message you reference looks like a defect in the hip package, which tries to invoke a dependency that is not being properly enforced by the concretizer.

I don't have convenient access to a ROCm@6.1 install I trust, but I'm able to build `gasnet` (and `upcxx`) just fine using ROCm@6.0 on Frontier with spack:develop 93ccd81d86:
```
frontier$ module list

Currently Loaded Modules:
  1) craype-x86-trento                       8) craype/2.7.19          15) lfs-wrapper/0.0.1
  2) libfabric/1.15.2.0                      9) cray-dsmml/0.2.2       16) DefApps/default
  3) craype-network-ofi                     10) cray-mpich/8.1.23      17) ums/default
  4) perftools-base/22.12.0                 11) cray-libsci/22.12.1.1  18) ums014/default
  5) xpmem/2.6.2-2.5_2.22__gd067c3f.shasta  12) PrgEnv-cray/8.3.3      19) upcxx-hip/nightly
  6) cray-pmi/6.1.8                         13) darshan-runtime/3.4.0  20) git/2.36.1
  7) cce/15.0.0                             14) hsi/default            21) rocm/6.0.0

frontier$ ./bin/spack spec gasnet+rocm conduits=ofi amdgpu_target=gfx90a ^hip@6.0
Input spec
--------------------------------
 -   gasnet+rocm amdgpu_target=gfx90a conduits=ofi
 -       ^hip@6.0

Concretized
--------------------------------
 -   gasnet@2024.5.0%gcc@12.2.0~cuda~debug~level_zero+rocm amdgpu_target=gfx90a build_system=generic conduits=ofi arch=linux-sles15-zen3
[+]      ^gcc-runtime@12.2.0%gcc@12.2.0 build_system=generic arch=linux-sles15-zen3
[e]      ^glibc@2.31%gcc@12.2.0 build_system=autotools arch=linux-sles15-zen3
[e]      ^hip@6.0.0%gcc@12.2.0~cuda+rocm build_system=cmake build_type=Release generator=make patches=1f65dfe,5bb9b0e,7668b2a,9165154,aee7249 arch=linux-sles15-zen3
[e]      ^hsa-rocr-dev@5.3.0%gcc@12.2.0+image+shared build_system=cmake build_type=Release generator=make patches=71e6851 arch=linux-sles15-zen3
[e]      ^libfabric@1.15.2%gcc@12.2.0~debug~kdreg build_system=autotools fabrics=cxi,sockets,tcp,udp arch=linux-sles15-zen3
[e]      ^llvm-amdgpu@5.3.0%gcc@12.2.0~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=a08bbe1 arch=linux-sles15-zen3

frontier$ ./bin/spack install --no-cache --overwrite --keep-stage --test=root -y -v gasnet+rocm conduits=ofi amdgpu_target=gfx90a ^hip@6.0
...
==> gasnet: Successfully installed gasnet-2024.5.0-sg7cduvmj7wnikfy3bolk2gv6ivkabpn
```

Strangely the concretization above includes `hsa-rocr-dev@5.3.0`, despite the fact that hip's package.py includes:
```
    with when("+rocm"):
        depends_on("gl@4.5:")
        depends_on("py-cppheaderparser", type="build", when="@5.3.3:")
        for ver in [
            "5.1.0",
            "5.1.3", 
            "5.2.0",
            "5.2.1",
            "5.2.3",
            "5.3.0",
            "5.3.3",
            "5.4.0",
            "5.4.3",
            "5.5.0",
            "5.5.1",
            "5.6.0",
            "5.6.1",
            "5.7.0",
            "5.7.1",
            "6.0.0",
            "6.0.2",
            "6.1.0",
            "6.1.1",
        ]:
            depends_on(f"hsakmt-roct@{ver}", when=f"@{ver}")
            depends_on(f"hsa-rocr-dev@{ver}", when=f"@{ver}")
```
So it seems like the concretizer is ignoring hip's expressed version dependency, which might indicate a deeper problem.. 
What does your concretization include for `hsa-rocr-dev`?

**Comment by bonachea** [comment]
FWIW, the [failing pipeline ](https://gitlab.spack.io/spack/spack/-/pipelines/722967) indicates this problem is in no way specific to `gasnet` - many other failing pipelines are dropping dead with the same error, eg: [kokkos](https://gitlab.spack.io/spack/spack/-/jobs/11464808) [petsc](https://gitlab.spack.io/spack/spack/-/jobs/11464920) [hiprand](https://gitlab.spack.io/spack/spack/-/jobs/11464927), etc.

So to clarify I guess I _should_ have said that I fixed the underlying problem in GASNet-EX sources that was preventing `gasnet` and `upcxx` from building against ROCm 6+. 

This other problem with the `hip` package not correctly resolving its `hsa-rocr-dev` dependency (in at least 6.1.1) apparently remains and affects 20+ packages (including `gasnet` and `upcxx`). I have no idea how to resolve that Spack problem, but it seems orthogonal to my recent work.

**Comment by eugeneswalker** [comment]
> This other problem with the `hip` package not correctly resolving its `hsa-rocr-dev` dependency (in at least 6.1.1) apparently remains and affects 20+ packages (including `gasnet` and `upcxx`). I have no idea how to resolve that Spack problem, but it seems orthogonal to my recent work.

Yes, exactly. This should be fixed here:
* https://github.com/spack/spack/pull/44528

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (tldahlgren) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Curious.  Is this the correct runner?

**Comment by bonachea** [comment]
@eugeneswalker Regarding the remaining failures on a few packages:
For at least [raja](https://gitlab.spack.io/spack/spack/-/jobs/11484740), [adios](https://gitlab.spack.io/spack/spack/-/jobs/11484820) and [umpire](https://gitlab.spack.io/spack/spack/-/jobs/11484817) failures that have this signature:
```
error: no member named 'memoryType' in 'hipPointerAttribute_t'
     131      114 |           switch (a.memoryType) {
     132          |                   ~ ^
```
This failure mode indicates the underlying package very likely has not been ported to ROCm 6.x, where the deprecated `memoryType` field of `hipPointerAttribute_t` was removed. This [documented breakage](https://rocm.docs.amd.com/en/latest/about/CHANGELOG.html#id106) in ROCm 6's backwards compatibility led to exactly the defect that we fixed in gasnet@2024.5.0.

So you may want to open defect reports to those package maintainers regarding that issue.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (eugeneswalker) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: No! Thanks for catching this. I just pushed a fix. Good eye!

**Review by tldahlgren — DISMISSED** [review]

**Comment by tldahlgren** [comment]
Still need to resolve the `hip` conflict resulting from merging #44528.

**Review by tldahlgren — DISMISSED** [review]

**Review by wspear — APPROVED** [review]

**Pull request merged**
